scope.py

```diff
@@ -0,0 +1,297 @@
+                pathlib.Path(__file__).parent.absolute() / "config.yaml"
+            )
+
+            # use token specified as env var (if exists)
+            kowalski_token_env = os.environ.get("KOWALSKI_TOKEN")
+            if kowalski_token_env is not None:
+                self.config["kowalski"]["token"] = kowalski_token_env
+
+        # try setting up K connection if token is available
+        if self.config["kowalski"]["token"] is not None:
+            with status("Setting up Kowalski connection"):
+                self.kowalski = Kowalski(
+                    token=self.config["kowalski"]["token"],
+                    protocol=self.config["kowalski"]["protocol"],
+                    host=self.config["kowalski"]["host"],
+                    port=self.config["kowalski"]["port"],
+                )
+        else:
+            self.kowalski = None
+            # raise ConnectionError("Could not connect to Kowalski.")
+            print("Kowalski not available")
+
+    def _get_nearest_gaia(
+            self,
+            positions: Sequence[Sequence[float]],
+            catalog: str = "Gaia_EDR3",
+            max_distance: float = 5.0,
+            distance_units: str = "arcsec",
+    ) -> pd.DataFrame:
+        """Get nearest Gaia source for a set of given positions
+
+        :param positions: R.A./Decl. [deg]
+        :param catalog: Gaia catalog to query
+        :param max_distance:
+        :param distance_units: arcsec | arcmin | deg | rad
+        :return:
+        """
+        if self.kowalski is None:
+            raise ConnectionError("Kowalski connection not established.")
+        query = {
+            "query_type": "near",
+            "query": {
+                "max_distance": max_distance,
+                "distance_units": distance_units,
+                "radec": positions,
+                "catalogs": {
+                    catalog: {
+                        "filter": {},
+                        "projection": {
+                            "parallax": 1,
+                            "parallax_error": 1,
+                            "pmra": 1,
+                            "pmra_error": 1,
+                            "pmdec": 1,
+                            "pmdec_error": 1,
+                            "phot_g_mean_mag": 1,
+                            "phot_bp_mean_mag": 1,
+                            "phot_rp_mean_mag": 1,
+                            "ra": 1,
+                            "dec": 1,
+                        }
+                    }
+                }
+            },
+            "kwargs": {
+                "limit": 1
+            }
+        }
+        response = self.kowalski.query(query=query)
+        gaia_nearest = [
+            v[0] for k, v in response.get("data").get(catalog).items()
+            if len(v) > 0
+        ]
+        df = pd.DataFrame.from_records(gaia_nearest)
+
+        df["M"] = df["phot_g_mean_mag"] + 5 * np.log10(df["parallax"] * 0.001) + 5
```

Comment:
The parallax can be <0 here, which results in a warning. Not sure if we want to catch that. It will return a Nan value, which does not get plotted, so its not catastrophic

Reply:
I'd keep the warnings -- the user will see those when building the docs, so can take action if necessary.

Reply:
Hmm. But what should the user do when these are seen? Remove those specific ones? Will that add bias?

Reply:
I think as long as there's at least one valid value here, should be fine - other stuff will just be ignored.